[Outlook] (manifest) new preview JSON manifest (#3428)
OfficeDev/office-js-docs-pr: docs/develop/json-manifest-overview.md

  ```json
  "items": [
      {
          -- markup for a menu item is here --
      },
      {
          -- markup for another menu item is here --
      }
  ]
  ```

#### Top-level structure

The root level of the preview JSON manifest, which roughly corresponds to the **OfficeApp** element in the current XML manifest, is an anonymous object. 

The children of **OfficeApp** are commonly divided into two notional categories. The **VersionOverrides** element is one category. The other consists of all the other children of **OfficeApp**, which are collectively referred to as the base manifest. So too, the preview JSON manifest has a similar division. There is a top-level "extension" property that roughly corresponds in its purposes and child properties to the **VersionOverrides** element. The preview JSON manifest also has over 10 other top-level properties that collectively serve the same purposes as the base manifest of the XML manifest. These other properties can be thought of collectively as the base manifest of the JSON manifest. 

> [!NOTE]

> > When it becomes possible combine an add-in with other Microsoft 365 extension types in a single manifest, then there will be other top-level properties that don't fit into the notion of the base manifest. There will usually be a top-level property for every kind of Microsoft 365 extension type, such as "configurableTabs", "bots" and "connectors". For examples, see the [Teams manifest documentation](/microsoftteams/platform/resources/schema/manifest-schema). This structure makes clear that the "extension" property represents an Office add-in as one type of Microsoft 365 extension.

Reviewer: Can you add a comma or the word "to" after the phrase "When it becomes possible?" I'm a bit confused to the meaning here.

#### Base manifest

> The base manifest properties specify characteristics of the add-in that *any* type of extension of Microsoft 365 would be expected to have. This includes Teams tabs and message extensions, not just Office add-ins. These characteristics include a public name and a unique ID. The following table shows a mapping of some critical top-level properties in the preview JSON manifest to the XML elements in the current manifest, where the mapping principle is the *purpose* of the markup.

Author: ```suggestion
The base manifest properties specify characteristics of the add-in that *any* type of extension of Microsoft 365 is expected to have. This includes Teams tabs and message extensions, not just Office add-ins. These characteristics include a public name and a unique ID. The following table shows a mapping of some critical top-level properties in the preview JSON manifest to the XML elements in the current manifest, where the mapping principle is the *purpose* of the markup.
```

---
title: Teams manifest for Office Add-ins (preview)
description: Get an overview of the preview JSON manifest.
ms.date: 05/24/2022
ms.localizationpriority: high
---

# Teams manifest for Office Add-ins (preview)

Microsoft is making a number of improvements to the Microsoft 365 developer platform. These improvements provide more consistency in the development, deployment, installation, and administration of all types of extensions of Microsoft 365, including Office Add-ins. These changes are compatible with existing add-ins. 

> Two important features we are aiming at, *but are not yet available*:

Reviewer 2: Can we update the to focus on the shared manifest/schema to enable a single unit of distribution - it's more aligned to the scope of the preview.  We can add the other point as needed at a subsequent milestone. 

For example:
One important improvement we are working on is the ability to create a single unit of distribution for all your Microsoft 365 extensions by using the same shared manifest format and schema, based on the current JSON formatted Teams manifest schema.  

We have taken a first step towards....

- It should be possible to surface a single web app as multiple types of Microsoft 365 extensions. For example, a web app can be both an Office Add-in and a custom tab in Teams.
- All types of Microsoft 365 extensions should use the same manifest format and schema. It will be based on the current JSON-formatted Teams manifest schema. 

We have taken an important first step toward these goals by making it possible for Outlook Add-ins, running on Windows only, to be created with a version of the Teams JSON manifest.

> [!NOTE]
> The new manifest is available for preview and is subject to change based on feedback. We encourage experienced add-in developers to experiment with it. The preview manifest should not be used in production add-ins. 

During the early preview period, the following limitations apply.

- The preview version of the Teams manifest only supports Outlook add-ins and only on subscription Office for Windows. We're working on extending support to Excel, PowerPoint, and Word.

> - It is not yet possible to combine an add-in with Teams apps, such as a custom Teams tab, or any other Microsoft 365 extension types. So, you cannot yet add an add-in to an existing Teams app. This will be available before release to general availability.

Reviewer 2: "It is not yet possible to combine and sideload an add-in with Teams apps, such as a Teams personal tab, or other Microsoft 365 extension types.  In the coming months, we will continue to extend the preview to support these scenarios and provide additional tools to update manifests to the preview format."

  ```json
  "items": [
      {
          -- markup for a menu item is here --
      },
      {
          -- markup for another menu item is here --
      }
  ]
  ```

#### Top-level structure

The root level of the preview JSON manifest, which roughly corresponds to the **OfficeApp** element in the current XML manifest, is an anonymous object. 

The children of **OfficeApp** are commonly divided into two notional categories. The **VersionOverrides** element is one category. The other consists of all the other children of **OfficeApp**, which are collectively referred to as the base manifest. So too, the preview JSON manifest has a similar division. There is a top-level "extension" property that roughly corresponds in its purposes and child properties to the **VersionOverrides** element. The preview JSON manifest also has over 10 other top-level properties that collectively serve the same purposes as the base manifest of the XML manifest. These other properties can be thought of collectively as the base manifest of the JSON manifest. 

> [!NOTE]

> > When it becomes possible to combine an add-in with other Microsoft 365 extension types in a single web app, then there will be other top-level properties that don't fit into the notion of the base manifest. There will usually be a top-level property for every kind of Microsoft 365 extension type, such as "configurableTabs", "bots" and "connectors". For examples, see the [Teams manifest documentation](/microsoftteams/platform/resources/schema/manifest-schema). This structure makes clear that the "extension" property represents an Office add-in as one type of Microsoft 365 extension.

Reviewer 2: "When you combine an add-in with other Microsoft 365 extension types in a single manifest, there will be... "

Suggest updating the "single web app" to "single manifest".  
We've noted that combining and sideloading isn't supported earlier in the doc, so I don't know if we need to call it out again here?

        "name": "Microsoft Corp.",
        "websiteUrl": "https://aka.ms/opc_metaos_examples",
        "privacyUrl": "https://aka.ms/opc_metaos_privacy",
        "termsOfUseUrl": "https://aka.ms/opc_metaos_examples"
    },
    "localizationInfo": {
        "defaultLanguageTag": "en-us",
        "additionalLanguages": [
            {
                "languageTag": "es-es",
                "file": "locales/es-es.json"
            }
        ]
    },
    "webApplicationInfo": {
        "id": "c62f9f19-d901-48c8-a184-9a69d83305bc",
        "resource": "api://www.oepbuild2017.com/prodapp"
    },
    "authorization": {
        "permissions": {
            "resourceSpecific": [
                {
                    "name": "Mailbox.ReadWrite",
                    "type": "Delegated"
                },
                {
                    "name": "Document.ReadWrite",
                    "type": "Delegated"
                }
            ]
        }

    },
    "extension": {
        "requirements": {
            "scopes": [
                "document",
                "presentation",
                "workbook",
                "mail"
            ],
            "capabilities": [

>                { "name": "AddinCommands", "minVersion": "1.1" }

Reviewer 3: Can we change to:
```
"capabilities": [
  { "name": "AddinCommands", "minVersion": "1.1" },
  { "name": "MailBox", "minVersion": "1.3" }
]
```
As we need the mailbox capability?

            },
    
            {
                "requirements": {
                    "capabilities": [
                        { "name": "MailBox", "minVersion": "1.10" }
                    ]
                },
                "id": "text",
                "type": "general",
                "code": {
                    "page": "https://aka.ms/opc_metaos_examples/alpha/elements/en-us/actions_text.html",
                    "script": "https://aka.ms/opc_metaos_examples/alpha/elements/en-us/actions_text.js"
                },
                "lifetime": "short",
                "actions": {
                    "items": [
                        {
                            "id": "text.onMessageSending",
                            "type": "execution",
                            "name": "OnMessageSending"
                        },
                        {
                            "id": "text.onNewMessageComposeCreated",
                            "type": "execution",
                            "name": "OnNewMessageComposeCreated"
                        }
                    ]
                }
            },
            {
                "requirements": {

>                     "capabilities": [

Reviewer 3: Can also remove this section related to CFs

                                        },
                                        "actionId": "text.toLowercase"
                                    }
                                ]
                            },
                            {
                                "id": "showDashboard",
                                "type": "button",
                                "label": "Show dashboard",
                                "icons": [
                                    { "size": 16, "file": "test_16.png" },
                                    { "size": 32, "file": "test_32.png" },
                                    { "size": 80, "file": "test_80.png" }
                                ],
                                "supertip": {
                                    "title": "Show dashboard",
                                    "description": "click to open dashboard"
                                },
                                "actionId": "text.showDashboard"
                            }
                        ]
                    }
                ]
            }
        ],
        "ribbons": [
            {
                "requirements": {

>                     "scopes": ["presentation"]

Reviewer 3: We can remove the one related to presentation

                                    }
                                ]
                            }
                        ]
                    }
                ]
            }
        ],
        "keyboards": [
            {
                "requirements": {
                    "capabilities": [
                        { "name": "SharedRuntime", "minVersion": "1.1" }
                    ],
                    "platforms": ["windows", "web"]
                },
                "shortcuts": [
                    {
                        "key":  "Ctrl+Shift+U",
                        "actionId": "text.toUppercase"
                    },
                    {
                        "key": "Ctrl+Shift+L",
                        "actionId": "text.toLowercase"
                    },
                    {
                        "key": "Ctrl+Shift+Up",
                        "actionId": "text.showDashboard"
                    }
                ]
            },
            {
                "requirements": {
                    "capabilities": [
                        { "name": "SharedRuntime", "minVersion": "1.1" }
                    ],

>                     "platforms": ["mac"]

Reviewer 3: Right now it doesn't work easily on mac, so I think we can take this part out.
The difference for mac is that installing `.net` dependency is much harder and we won't help customers with that in the tooling

                    ]
                },
                "events": [
                    {
                        "id": "newMessageComposeCreated",
                        "actionId": "text.onNewMessageComposeCreated"
                    },
                    {
                        "id": "messageSending",
                        "actionId": "text.onMessageSending",
                        "options": {
                            "sendMode": "promptUser"
                        }
                    }
                ]
            }
        ],
        "alternatives": [
            {
                "requirements": {
                    "scopes": ["mail"]
                },
                "prefer": {
                    "comAddin": {
                        "progId": "ContosoExtension"
                    }
                },
                "hide": {
                    "storeOfficeAddin": {
                        "officeAddinId": "fca2794d-4aa5-4023-a84b-c60a3cbd33d4",
                        "assetId": "WA129485"
                    }
                }
            },
            {
                "requirements": {

>                     "scopes": ["presentation", "document", "workbook"]

Reviewer 3: Take this out

Two of the most important improvements are:

- It will be possible to surface a single web app as multiple types of Microsoft 365 extensions. For example a web app can be both an Office Add-in and a custom tab in Teams.
- All types of Microsoft 365 extensions will use the same manifest format (JSON) and schema. It will be based on the current Teams manifest schema. In support of the first bullet, it will be possible to specify multiple types of extensions in the manifest.  

The new manifest is available for preview and we encourage experienced add-in developers to experiment with it. It should not be used in production add-ins. During the early preview period, the following limitations apply:

- The preview version of the Teams manifest only supports Outlook add-ins and only on subscription Office for Windows. We're working on extending support to Excel, PowerPoint, and Word.
- It is not yet possible to combine an add-in with Teams apps, such as a custom Teams tab, or any other Microsoft 365 extension types. We're working on this too.

> [!TIP]

> 

Reviewer 4: Extra blank line.
```suggestion
> [!TIP]
```

- JSON does not have the XML distinction between attribute and element value. So typically the JSON that maps to an XML element makes both the element value and each of the attributes a child property. For example, the following shows some XML markup and its JSON equivalent.
  
  ```xml
  <MyThing color="blue">Some text</MyThing>
  ```

  ```json
  "myThing" : {
      "color": "blue",
      "text": "Some text"
  }
  ```
- There are many places in the current XML manifest where an element with a plural name has children with the singular version of the same name. For example, the markup to configure a custom menu includes an **Items** element which can have multiple **Item** element children. The JSON equivalent of these plural elements is a property with an array as its value. The members of the array are *anonymous* objects, not properties named "item" or "item1", "item2", etc. The following is an example.

  ```json
  "items": [
      {
          -- markup for a menu item is here --
      },
      {
          -- markup for another menu item is here --
      }
  ]
  ```

> #### Top level structure

Reviewer 4: I _think_ top-level should be hyphenated here.
```suggestion
#### Top-level structure
```

The root level of the preview JSON manifest, which roughly corresponds to the **OfficeApp** element in the current XML manifest, is an anonymous object. 

The children of **OfficeApp** are commonly divided into two notional categories. The **VersionOverrides** element is one category. The other consists of all the other children of **OfficeApp** which are collectively referred to as the base manifest. So too, the preview JSON manifest has a similar division. There is a top-level "extension" property that roughly corresponds in its purposes and child properties to the **VersionOverrides** element. The preview JSON manifest also has over 10 other top-level properties that collectively serve the same purposes as the base manifest of the XML manifest. So, these other properties can be thought of collectively as the base manifest of the JSON manifest. 

> [!NOTE]

> > When it becomes possible to combine an add-in with other Microsoft 365 extension types in a single web app, then there will be other top-level properties that don't fit into the notion of the base manifest. Roughly speaking, there will be a top-level property for every kind of Microsoft 365 extension type, such as "configurableTabs", "bots" and "connectors". For examples, see the [Teams manifest documentation](/microsoftteams/platform/resources/schema/manifest-schema). This structure makes clear that the "extension" property represents an Office add-in as one type of Microsoft 365 extension.

Reviewer 4: Phrasing change for ease of localization.
```suggestion
> When it becomes possible to combine an add-in with other Microsoft 365 extension types in a single web app, then there will be other top-level properties that don't fit into the notion of the base manifest. There will usually be a top-level property for every kind of Microsoft 365 extension type, such as "configurableTabs", "bots" and "connectors". For examples, see the [Teams manifest documentation](/microsoftteams/platform/resources/schema/manifest-schema). This structure makes clear that the "extension" property represents an Office add-in as one type of Microsoft 365 extension.
```

#### Base manifest

> The base manifest properties specify characteristics of the add-in that *any* type of extension of Microsoft 365, including Teams tabs or message extensions, would be expected to have, not just Office add-ins. These characteristics include a public name and a unique ID. The following table shows a mapping of some critical top level properties in the preview JSON manifest to the XML elements in the current manifest, where the mapping principle is the *purpose* of the markup.

Reviewer 4: Consider splitting the first sentence into two for readability. 
```suggestion
The base manifest properties specify characteristics of the add-in that *any* type of extension of Microsoft 365 would be expected to have. This includes Teams tabs and message extensions, not just Office add-ins. These characteristics include a public name and a unique ID. The following table shows a mapping of some critical top-level properties in the preview JSON manifest to the XML elements in the current manifest, where the mapping principle is the *purpose* of the markup.
```

> |JSON property|Purpose|XML element(s)|Comments|

Reviewer 4: Are all of these required? If so, consider putting that in the preceding paragraph. If not, consider adding a column to the table specifying which ones are required.

Author: They are not all required, but I don't think we should get so deep into the weeds that we are specifying what is required. This isn't reference documentation. It just orients someone familiar with the XML manifest to the JSON manifest.

|:-----|:-----|:-----|:-----|
|"$schema"| Identifies the manifest schema. | attributes of **OfficeApp** and **VersionOverrides** | |
|"id"| GUID of the add-in. | **Id**| |
|"version"| Version of the add-in. | **Version** | |
|"manifestVersion"| Version of the manifest schema. |  attributes of **OfficeApp** | |
|"name"| Public name of the add-in. | **DisplayName** | |
|"description"| Public description of the add-in.  | **Description** | |

> |"accentColor"||| This property has no equivalent in the current XML manifest and is not used in the preview of the JSON manifest. But it must be present. |

Reviewer 4: For me, the takeaway is that you need to add this, but it does nothing. I don't think more needs to be said.
```suggestion
|"accentColor"||| Required, but isn't supported in the preview. |
```

Author: I prefer it as is. This table is meant to orient someone familiar with the XML manifest, by loosely mapping elements to properties. It is not reference documentation for the properties.

|"developer"| Identifies the developer of the add-in. | **ProviderName** | |
|"localizationInfo"| Configures the default locale and other supported locales. | **DefaultLocale** and **Override** | |
|"webApplicationInfo"| Identifies the add-in's web app as it is known in Azure Active Directory. | **WebApplicationInfo** | In the current XML manifest, the **WebApplicationInfo** element is inside **VersionOverrides**, not the base manifest. |

> |"authorization"| Identifies any Microsoft Graph permissions that the add-in needs. | **WebApplicationInfo** | See comment in preceding row. |

Reviewer 4: Might as well spell this out, in case someone changes the preceding row without reading the next one.
```suggestion
|"authorization"| Identifies any Microsoft Graph permissions that the add-in needs. | **WebApplicationInfo** | In the current XML manifest, the **WebApplicationInfo** element is inside **VersionOverrides**, not the base manifest. |
```

> The **Hosts**, **Requirements**, and **ExtendedOverrides** elements are part of the base manifest in the current XML manifest. But concepts and purposes associated with these elements are configured inside the "extension" property of the preview JSON manifest. 

Reviewer 4: Does the concept of a "base manifest" exist in the XML version today?

Author: Yes. For an example, see https://docs.microsoft.com/en-us/office/dev/add-ins/develop/specify-office-hosts-and-api-requirements

#### "extension" property

> The "extension" property in the preview JSON manifest primarily represents characteristics of the add-in that would not be relevant to other kinds of Microsoft 365 extensions. For example, the Office applications that the add-in extends (such as, Excel, PowerPoint, Word, and Outlook) are specified inside the "extension" property as are customizations of the Office application ribbon. The property's purposes closely match those of the **VersionOverrides** element in the current XML manifest.

Reviewer 4: Comma suggestion. Also, something about the phrase "The property's purposes" seems off to me, but I'm not sure why. 

```suggestion
The "extension" property in the preview JSON manifest primarily represents characteristics of the add-in that would not be relevant to other kinds of Microsoft 365 extensions. For example, the Office applications that the add-in extends (such as, Excel, PowerPoint, Word, and Outlook) are specified inside the "extension" property, as are customizations of the Office application ribbon. The property's purposes closely match those of the **VersionOverrides** element in the current XML manifest.
```

Author: ```suggestion
The "extension" property in the preview JSON manifest primarily represents characteristics of the add-in that would not be relevant to other kinds of Microsoft 365 extensions. For example, the Office applications that the add-in extends (such as, Excel, PowerPoint, Word, and Outlook) are specified inside the "extension" property, as are customizations of the Office application ribbon. The configuration purposes of the "extension" property closely match those of the **VersionOverrides** element in the current XML manifest.
```

> [!NOTE]

> > The **VersionOverrides** section of the manifest has a "double jump" system for many string resources. Strings, including URLs, are specified and assigned an ID in the **Resources** child of **VersionOverrides**. Elements that require a string have a `resid` attribute that matches the ID of a string in the **Resources** element. The "extension" property of the preview JSON manifest simplifies things by defining strings directly as property values. There is nothing equivalent to the **Resources** element.

Reviewer 4: Is this note about the current manifest?

Author: ```suggestion
> The **VersionOverrides** section of the current XML manifest has a "double jump" system for many string resources. Strings, including URLs, are specified and assigned an ID in the **Resources** child of **VersionOverrides**. Elements that require a string have a `resid` attribute that matches the ID of a string in the **Resources** element. The "extension" property of the preview JSON manifest simplifies things by defining strings directly as property values. There is nothing in the JSON manifest that is equivalent to the **Resources** element.
```

##### "ribbons" table

The following table maps the child properties of the anonymous child objects in the "ribbons" array onto XML elements in the current manifest. 

|JSON property|Purpose|XML element(s)|Comments|
|:-----|:-----|:-----|:-----|
| "contexts" | Specifies the command surfaces that the add-in customizes. | various **\*CommandSurface** elements, such as **PrimaryCommandSurface** and **MessageReadCommandSurface** |  |
| "tabs" | Configures custom ribbon tabs. | **CustomTab** | The names and hierarchy of the descendant properties of "tabs" closely match the descendants of **CustomTab**.  |

## Sample preview JSON manifest

The following is an example of a preview JSON-manifest for an add-in.

```json
{
    "$schema": "../../schema/metaos.public.schema.json",
    "id": "82a9d9c3-4702-4322-bbc4-6fe7f9b01483",
    "version": "1.0.0",
    "manifestVersion": "devPreview",
    "name": {
        "short": "Basic Office Example",
        "full": "Transform text to uppercase/lowercase."
    },
    "description": {
        "short": "Example MetaOS app that demonstrates various Office features.",

>         "full": "Example MetaOS app that demonstrates various Office features like ribbon, menubar, context menu, keyboard shortcuts, custom functions."

Reviewer 4: I thought "MetaOS" wasn't a public term?

    },
    "icons": {
        "outline": "small_icon.png",
        "color": "color_icon.png"
    },
    "accentColor": "#230201",
    "developer": {
        "name": "Microsoft Corp.",
        "websiteUrl": "https://aka.ms/opc_metaos_examples",
        "privacyUrl": "https://aka.ms/opc_metaos_privacy",

>         "termsOfUseUrl": "https://aka.ms/opc_metaos_examples"

Reviewer 4: These URLs don't go anywhere for me.

---
title: Teams manifest for Office Add-ins (preview)
description: Get an overview of the preview JSON manifest.
ms.date: 05/24/2022
ms.localizationpriority: high
---

# Teams manifest for Office Add-ins (preview)

> 

Author: We already have this near the top of both articles: "The new manifest is available for preview and we encourage experienced add-in developers to experiment with it. It should not be used in production add-ins."

I will add the line about subject to change and make it a note.

The children of **OfficeApp** are commonly divided into two notional categories. The **VersionOverrides** element is one category. The other consists of all the other children of **OfficeApp**, which are collectively referred to as the base manifest. So too, the preview JSON manifest has a similar division. There is a top-level "extension" property that roughly corresponds in its purposes and child properties to the **VersionOverrides** element. The preview JSON manifest also has over 10 other top-level properties that collectively serve the same purposes as the base manifest of the XML manifest. These other properties can be thought of collectively as the base manifest of the JSON manifest. 

> [!NOTE]

> > When it becomes possible combine an add-in with other Microsoft 365 extension types in a single manifest, then there will be other top-level properties that don't fit into the notion of the base manifest. There will usually be a top-level property for every kind of Microsoft 365 extension type, such as "configurableTabs", "bots" and "connectors". For examples, see the [Teams manifest documentation](/microsoftteams/platform/resources/schema/manifest-schema). This structure makes clear that the "extension" property represents an Office add-in as one type of Microsoft 365 extension.

Reviewer 2: was removing the "to" here intentional?

Author: Good catch.
```suggestion
> When it becomes possible to combine an add-in with other Microsoft 365 extension types in a single manifest, then there will be other top-level properties that don't fit into the notion of the base manifest. There will usually be a top-level property for every kind of Microsoft 365 extension type, such as "configurableTabs", "bots" and "connectors". For examples, see the [Teams manifest documentation](/microsoftteams/platform/resources/schema/manifest-schema). This structure makes clear that the "extension" property represents an Office add-in as one type of Microsoft 365 extension.
```
